File: 백준/Silver/1912. 연속합/연속합.py
import sys

input = sys.stdin.readline

N = int(input())

nums = list(map(int, input().split()))

dp = [0] * N
dp[0] = nums[0]
maxNum = dp[0]

for i in range(1, N):
    dp[i] = max(dp[i-1] + nums[i], nums[i])
    maxNum = max(maxNum, dp[i])

print(maxNum)

File: 백준/Silver/1912. 연속합/연속합.swift
import Foundation

let n = Int(readLine()!)!
let nums = readLine()!.split(separator: " ").map { Int(String($0))! }

var memo = Array(repeating: Int.min, count: n)

func recursive(idx: Int) -> Int {
  if idx == n { return 0 }

  if memo[idx] != Int.min { return memo[idx] }

  let result = recursive(idx: idx + 1)

  memo[idx] = max(nums[idx] + result, nums[idx])
  return memo[idx]
}

_ = recursive(idx: 0)
print(memo.max()!)